android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivityTest.kt

```diff
@@ -197,6 +205,67 @@ class QuestionnaireActivityTest : RobolectricTest() {
     Assert.assertNotNull(dialog)
   }
 
+  @Test
+  fun `setupLocationServices should fetch location when location is enabled and permissions granted`() {
+    val viewModel = mockk<QuestionnaireViewModel>()
+    val activity = spyk(QuestionnaireActivity())
+    val fusedLocationProviderClient = mockk<FusedLocationProviderClient>()
+    every { viewModel.applicationConfiguration.logQuestionnaireLocation } returns true
+    every { LocationServices.getFusedLocationProviderClient(activity) } returns
+      fusedLocationProviderClient
+    every { LocationUtils.isLocationEnabled(activity) } returns true
+    every { activity.hasLocationPermissions() } returns true
+    every { activity.fetchLocation(any()) } just runs
+
+    activity.setupLocationServices()
+
+    verify(exactly = 1) { activity.fetchLocation(true) }
+    verify(exactly = 0) { activity.openLocationServicesSettings() }
+    verify(exactly = 0) { activity.launchLocationPermissionsDialog() }
+  }
+
+  @Test
+  fun `setupLocationServices should open location settings if location is disabled`() {
+    val viewModel = mockk<QuestionnaireViewModel>()
+    val activity = spyk(QuestionnaireActivity())
+    val fusedLocationProviderClient = mockk<FusedLocationProviderClient>()
+    every { viewModel.applicationConfiguration.logQuestionnaireLocation } returns true
+    every { LocationServices.getFusedLocationProviderClient(activity) } returns
+      fusedLocationProviderClient
+    every { LocationUtils.isLocationEnabled(activity) } returns false
+    every { activity.hasLocationPermissions() } returns true
+    every { activity.fetchLocation(any()) } just runs
+    every { activity.openLocationServicesSettings() } just runs
+    every { activity.launchLocationPermissionsDialog() } just runs
```

Comment:
Can we simulate the actual behavior of the activity class instead of using mocks? You can launch the activity using the `ActivityLauncher` or `Robolectric` classes.